### PR TITLE
Add Examples for How to write Terraform Modules Article Part 1

### PR DESCRIPTION
**Pull request by mariux**
TODO:
- finalize links to final article

### CODE REVIEW

File: terraform/examples/module_depends_on/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+### Cloning the repository
+```
+git clone https://github.com/mineiros-io/article-examples.git
+cd article-examples/terraform/examples/module_depends_on
+```
+
+### Initializing Terraform
+Run `terraform init` to initialize the example. The output should look like:
+```
+Initializing modules...
```
Comment (soerenmartius) [inline terraform/examples/module_depends_on/README.md:59]: You could also work with gifs for showing example output. [Terminalizer](https://github.com/faressoft/terminalizer) is a great tool for creating gifs.

File: terraform/examples/module_depends_on/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+should now work.
+
+If you ever set or change modules or backend configuration for Terraform,
+rerun this command to reinitialize your working directory. If you forget, other
+commands will detect it and remind you to do so if necessary.
+```
+
+### Applying the example
+Run `terraform apply -auto-approve` to create the resources.
+You can clearly see in the output how creation of
```
Comment (soerenmartius) [inline terraform/examples/module_depends_on/README.md]: We should avoid expressions such as clearly. Let's try to be on point only :)

File: terraform/examples/module_depends_on/main.tf [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+module "has-no-dependency" {
+  source  = "../../modules/terraform-null-resources"
+
+  string = "Whenever!"
+}
+
+# will be created/destroyed as soon as possible
+resource "null_resource" "is-no-dependency" {}
+
+## make use of module_depends_on to define hidden external dependencies
```
Comment (soerenmartius) [inline terraform/examples/module_depends_on/main.tf]: Let's use the same comment syntax that we also use for our [modules](https://github.com/mineiros-io/terraform-module-template/blob/master/modules/vpc/main.tf#L11)?

File: terraform/examples/module_enabled/README.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
+
+# Example for the `module_enabled` feature
```
Comment (soerenmartius) [inline terraform/examples/module_enabled/README.md]: Let's use something more expressive here, e.g. `How to dynamically enable modules using the module_enabled toggle.`

I also believe that we can shorten syntax, e.G. `enable = true`

File: terraform/modules/terraform-null-resources/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  source  = "./base-module-full_module_configuration"
+
+  list_of_strings = ["hello", "world", "!"]
+
+  module_depends_on = [
+    null_resource.dependency
+  ]
+}
+
+# will be created before all module.example.* resources
```
Comment (soerenmartius) [inline terraform/modules/terraform-null-resources/README.md]: align with our comment standard

File: terraform/modules/terraform-null-resources/main.tf [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  depends_on = [
+    var.module_depends_on
+  ]
+}
+
+# create one resource for every element in var.list_of_strings
+# but only if var.module_enabled is true
+# resources will be created as a list of objects
+# if the var.list_of_strings is empty no resources will be created
```
Comment (soerenmartius) [inline terraform/modules/terraform-null-resources/main.tf]: multiline comments should follow our standard and represent whole sentences

File: terraform/modules/terraform-null-resources/outputs.tf [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+output "list_of_strings" {
+  description = "A list of strings"
+  value       = var.list_of_strings
+}
+
+# output all resources as-is
+output "single" {
+  description = "The single resource object"
+  value       = try(null_resource.single[0], {})
+  # try() was fully implemented in terraform 0.12.20
```
Comment (soerenmartius) [inline terraform/modules/terraform-null-resources/outputs.tf]: imho we don't need to explain that

**Review by soerenmartius — CHANGES_REQUESTED** [review]

File: terraform/examples/module_enabled/README.md (hunk 4 above)
Comment (mariux) [reply terraform/examples/module_enabled/README.md]: shorter syntax might conflict with resource `enabled` fields... since this is actually just a workaround for missing terraform features i would avoid any potential clash in naming by prefixing..

File: terraform/examples/module_depends_on/README.md (hunk 1 above)
Comment (mariux) [reply terraform/examples/module_depends_on/README.md:59]: yeah for initialize this makes sense... but for the other examples it is hard to follow details in animations... so i would rather keep it aligned..

File: terraform/examples/module_depends_on/README.md (hunk 2 above)
Comment (mariux) [reply terraform/examples/module_depends_on/README.md]: fixed

File: terraform/examples/module_depends_on/main.tf (hunk 3 above)
Comment (mariux) [reply terraform/examples/module_depends_on/main.tf]: fixed

File: terraform/examples/module_enabled/README.md (hunk 4 above)
Comment (mariux) [reply terraform/examples/module_enabled/README.md]: fixed heading

File: terraform/modules/terraform-null-resources/README.md (hunk 5 above)
Comment (mariux) [reply terraform/modules/terraform-null-resources/README.md]: fixed

File: terraform/modules/terraform-null-resources/main.tf (hunk 6 above)
Comment (mariux) [reply terraform/modules/terraform-null-resources/main.tf]: fixed

File: terraform/modules/terraform-null-resources/outputs.tf (hunk 7 above)
Comment (mariux) [reply terraform/modules/terraform-null-resources/outputs.tf]: removed

**Comment by mariux** [comment]
@soerenmartius adressed your comments .. some open discussions about naming and animations.. 
updated initial comment with TODO list of open topics..

**Review by soerenmartius — APPROVED** [review]
lgtm
